### PR TITLE
i#2511: isolate raw2trace

### LINKED ISSUE #2511: Make raw2trace more isolated
Right now, it aborts on error and uses global options. Fix that.

Also, it does a lot of directory walking and raw IO, which doesn't work with more exotic filesystems, where we can be storing very large traces.
Split out a self-contained library, which:
- only takes already opened std::istream and std::ostream-compatible files and doesn't own them;
- takes in the contents of the module file to avoid doing IO from drmodtrack_offline_read();
- returns errors on failure;
- doesn't depend on dr_frontend.

Also, add a frontend helper, which preserves the current behavior for local use on regular POSIX and Windows.

### PR DESCRIPTION
**Pull request by s-kanev**
Splits out a self-contained library, which:
- only takes already opened std::istream and std::ostream-compatible files
and doesn't own them, so it can be used with more exotic filesystems;
- takes in the contents of the module file to avoid doing IO from
drmodtrack_offline_read();
- returns errors on failure;
- doesn't depend on dr_frontend.

Preserves the current behavior with raw2trace_helper_t, which walks
directories and opens files.

TESTED: ctest (failures failed before this commit as well)

Fixes #2511

### CODE REVIEW

**Comment by s-kanev** [comment]
AppVeyor seems to fail on installing doxygen. I don't have a Windows box handy, if you give me credentials on one, I can take a look.

**Comment by derekbruening** [comment]
This doxygen install issue appeared a few days ago out of the blue, I was hoping it was some transient failure that would just go away (initially it surfaced as more of a network not-found error) but it doesn't seem to be the case.

**Comment by fhahn** [comment]
I think choco might got upgraded and the new default now is to not allow empty checksums by default. Ideally, the doxygen package would come with a proper checksum, but for now, installing it with `--allow-empty-checksums` should fix the builds: #2513

**Comment by fhahn** [comment]
The appveyor builds should be fixed now, I've used the `Update branch` button, to trigger a new build.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 1]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
+        }
         if (size > 0) {
             CHECK((uint)size < MAX_COMBINED_ENTRIES, "Too many entries");
-            if (!out_file.write((char*)buf_base, size))
-                FATAL_ERROR("Failed to write to output file");
+            if (!out_file->write((char*)buf_base, size))
+                return "Failed to write to output file";
         }
     } while (thread_count > 0);
+    return std::string();
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: Sometimes you're using 'return ""' and others 'return std::string()' -- best to pick one.  IMHO "" is clearer.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 2]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
+    if (!error.empty())
+        return error;
     trace_entry_t entry;
     entry.type = TRACE_TYPE_HEADER;
     entry.size = 0;
     entry.addr = TRACE_ENTRY_VERSION;
-    if (!out_file.write((char*)&entry, sizeof(entry)))
-        FATAL_ERROR("Failed to write header to output file %s", outname.c_str());
+    if (!out_file->write((char*)&entry, sizeof(entry)))
+        return "Failed to write header to output file ";
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: Why the trailing space?  I suppose we don't have the name anymore.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 3]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
-    open_thread_files();
     merge_and_process_thread_files();
 
     entry.type = TRACE_TYPE_FOOTER;
     entry.size = 0;
     entry.addr = 0;
-    if (!out_file.write((char*)&entry, sizeof(entry)))
-        FATAL_ERROR("Failed to write footer to output file %s", outname.c_str());
+    if (!out_file->write((char*)&entry, sizeof(entry)))
+        return "Failed to write footer to output file ";
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: Ditto

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 4]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
+    return std::string();
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: See earlier comment vs ""

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 5]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
+            if (!out_file->write((char*)buf_base, size))
+                return "Failed to write to output file";
         }
     } while (thread_count > 0);
+    return std::string();
 }
 
-void
+std::string
+raw2trace_t::check_thread_file(std::istream* fi)
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: style: *f

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 6]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
+    return std::string();
 }
 
-raw2trace_t::raw2trace_t(std::string indir_in, std::string outname_in)
-    : indir(indir_in), outname(outname_in), prev_instr_was_rep_string(false),
-      instrs_are_separate(false)
+raw2trace_t::raw2trace_t(const char* module_map_in,
+                         const std::vector<std::istream*>& thread_files_in,
+                         std::ostream* out_file_in,
+                         void* dcontext_in,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: style 4x: type {*,&}var

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 7]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
-    if (indir.find(OUTFILE_SUBDIR) == std::string::npos)
-        indir += std::string(DIRSEP) + OUTFILE_SUBDIR;
-    out_file.open(outname.c_str(), std::ofstream::binary);
-    if (!out_file)
-        FATAL_ERROR("Failed to open output file %s", outname.c_str());
-    VPRINT(1, "Writing to %s\n", outname.c_str());
-
-    dcontext = dr_standalone_init();
+    if (dcontext == NULL) {
+      dcontext = dr_standalone_init();
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: style: indentation

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 8]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
 #ifdef ARM
-    // We keep the mode at ARM and rely on LSB=1 offsets in the modoffs fields
-    // to trigger Thumb decoding.
-    dr_set_isa_mode(dcontext, DR_ISA_ARM_A32, NULL);
+      // We keep the mode at ARM and rely on LSB=1 offsets in the modoffs fields
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: style: indentation

File: clients/drcachesim/tracer/raw2trace.h [hunk 9]
```diff
@@ -58,32 +58,36 @@ struct module_t {
 
 class raw2trace_t {
 public:
-    raw2trace_t(std::string indir, std::string outname);
+    // module_map, thread_files and out_file are all owned and opened/closed by the
+    // caller.
+    raw2trace_t(const char* module_map, const std::vector<std::istream*>& thread_files,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.h]: More pointer style, here and below

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 10]
```diff
@@ -535,68 +441,90 @@ raw2trace_t::merge_and_process_thread_files()
...
 }
 
-void
+std::string
+raw2trace_t::check_thread_file(std::istream* fi)
+{
+    // Check version header.
+    offline_entry_t ver_entry;
+    if (!fi->read((char*)&ver_entry, sizeof(ver_entry))) {
+        return "Unable to read thread log file ";
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: Trailing space again

File: clients/drcachesim/tracer/raw2trace_helper.cpp [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        return;
+    if (dr_snprintf(path, BUFFER_SIZE_ELEMENTS(path), "%s%s%s",
+                    indir.c_str(), DIRSEP, basename) <= 0) {
+        FATAL_ERROR("Failed to get full path of file %s", basename);
+    }
+    NULL_TERMINATE_BUFFER(path);
+    thread_files.push_back(new std::ifstream(path, std::ifstream::binary));
+    if (!(*thread_files.back()))
+        FATAL_ERROR("Failed to open thread log file %s", path);
+    raw2trace_t::check_thread_file(thread_files.back());
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Return value should be checked, and print out file name + error on failure

File: clients/drcachesim/tracer/raw2trace_helper.cpp [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* raw2trace helper to iterate directories and open files.
+ * Separate from raw2trace_t, so that raw2trace doesn't depend on dr_frontend. */
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: style: prefer */ on own line

File: clients/drcachesim/tracer/raw2trace_helper.cpp [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+#include "raw2trace_helper.h"
+
+#define FATAL_ERROR(msg, ...) do { \
+    fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__);    \
+    fflush(stderr); \
+    exit(1); \
+} while (0)
+
+#define CHECK(val, msg, ...) do { \
+      if (!(val)) FATAL_ERROR(msg, ##__VA_ARGS__); \
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: indent looks off

File: clients/drcachesim/tracer/raw2trace_helper.cpp [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+} while (0)
+
+#define VPRINT(level, ...) do { \
+      if (this->verbosity >= (level)) { \
+                fprintf(stderr, "[drmemtrace]: "); \
+                fprintf(stderr, __VA_ARGS__); \
+            } \
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: indent is off in above lines

File: clients/drcachesim/tracer/raw2trace_helper.cpp [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    NULL_TERMINATE_BUFFER(path);
+    thread_files.push_back(new std::ifstream(path, std::ifstream::binary));
+    if (!(*thread_files.back()))
+        FATAL_ERROR("Failed to open thread log file %s", path);
+    raw2trace_t::check_thread_file(thread_files.back());
+    VPRINT(1, "Opened thread log file %s\n", path);
+}
+
+raw2trace_helper_t::raw2trace_helper_t(const std::string& indir_in,
+                                       const std::string& outname_in,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: More style

File: clients/drcachesim/tracer/raw2trace_helper.h [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <fstream>
+#include <string>
+#include <vector>
+
+#include "dr_api.h"
+
+class raw2trace_helper_t {
+public:
+    raw2trace_helper_t(const std::string& indir, const std::string& outname,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_helper.h]: More style

**Comment by fhahn** [comment]
Can one of the admins verify this patch?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 17]
```diff
@@ -271,14 +163,14 @@ instr_is_rep_string(instr_t *instr)
 #endif
 }
 
-trace_entry_t *
-raw2trace_t::append_memref(trace_entry_t *buf_in, uint tidx, instr_t *instr,
+std::string
+raw2trace_t::append_memref(trace_entry_t **buf_in, uint tidx, instr_t *instr,
```
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 18]
```diff
@@ -308,12 +200,12 @@ raw2trace_t::append_memref(trace_entry_t *buf_in, uint tidx, instr_t *instr,
...
-    ++buf;
-    return buf;
+    *buf_in = ++buf;
+    return "";
 }
 
-bool
-raw2trace_t::append_bb_entries(uint tidx, offline_entry_t *in_entry)
+std::string
+raw2trace_t::append_bb_entries(uint tidx, offline_entry_t *in_entry, bool *handled)
```
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 1 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 5 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 10 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 2 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done. Sloppy edits.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 3 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 4 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 6 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 7 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 8 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.h (hunk 9 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace.h]: Done.

File: clients/drcachesim/tracer/raw2trace_helper.cpp (hunk 12 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace_helper.cpp (hunk 13 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace_helper.cpp (hunk 14 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace_helper.cpp (hunk 11 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Done. Thanks for catching.

File: clients/drcachesim/tracer/raw2trace_helper.cpp (hunk 15 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace_helper.h (hunk 16 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_helper.h]: Done.

File: clients/drcachesim/tracer/raw2trace_directory.h [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "dr_api.h"
+
+class raw2trace_directory_t {
+public:
+    raw2trace_directory_t(const std::string &indir, const std::string &outname,
+                          unsigned int verbosity = 0);
+    ~raw2trace_directory_t();
+
+    char* modfile_bytes;
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_directory.h]: One more style issue.

File: clients/drcachesim/tracer/raw2trace_launcher.cpp [hunk 20]
```diff
@@ -58,11 +52,16 @@ static droption_t<std::string> op_out
...
  "Specifies the path to the output file.");
 
-// Non-static for use by raw2trace.cpp
-droption_t<unsigned int> op_verbose
+static droption_t<unsigned int> op_verbose
 (DROPTION_SCOPE_FRONTEND, "verbose", 0, "Verbosity level for diagnostic output",
  "Verbosity level for diagnostic output.");
 
+#define FATAL_ERROR(msg, ...) do { \
+      fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__);    \
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace_launcher.cpp]: indentation: Looks like this is 6 spaces?

File: clients/drcachesim/tracer/raw2trace_directory.h (hunk 19 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_directory.h]: Done.

File: clients/drcachesim/tracer/raw2trace_launcher.cpp (hunk 20 above)
Comment (s-kanev) [inline clients/drcachesim/tracer/raw2trace_launcher.cpp]: Done.

**Comment by s-kanev** [comment]
> Could we add a test for the i/o interface?
Sure, added one through the drraw2trace tool. PTAL because I'm very bad at cmake.
There's also a slightly finer-grained test in this change's internal counterpart.

**Comment by derekbruening** [comment]
> > Could we add a test for the i/o interface?
> Sure, added one through the drraw2trace tool. PTAL because I'm very bad at cmake.

Hmm, I guess I was expecting a custom test of the interface for other than the built-in dir iterator, to ensure the interface fully isolates filesystem access: akin to the burst_replaceall test.

**Comment by derekbruening** [comment]
> > > Could we add a test for the i/o interface?
> 
> > Sure, added one through the drraw2trace tool. PTAL because I'm very bad at cmake.
> 
> Hmm, I guess I was expecting a custom test of the interface for other than the built-in dir iterator, to ensure the interface fully isolates filesystem access: akin to the burst_replaceall test.

I think my pointing at the existing end-to-end -indir tests was misleading: I meant that to say we already have a test of the dir iterator, not to suggest that a new test of using the interface for different code from the -indir should follow that template.

**Comment by s-kanev** [comment]
> I meant that to say we already have a test of the dir iterator

Gotcha, I didn't realize drcachesim's -indir exercises the iterator as well (through analyzer_multi_t), though in hindsight that's obvious. In this case, what I added is definitely redundant.
I'll add a standalone test that just invokes raw2trace_t on a set of raw files, similar to what we do downstream.

**Comment by s-kanev** [comment]
This took embarrassingly long. I couldn't use the file op replacement API like in burst_replaceall, because the code we're trying to trace itself calls drmodtrack_* and that didn't seem to isolate well (always ended up segfaulting). So, I ended up just doing a simple ptrace syscall recorder and checking that all new files that we open are only because of loading new modules.
We could get more sophisticated -- e.g. checking that all trace reads/writes are to pre-opened file descriptors -- but I don't think it's worth the extra effort.

File: clients/drcachesim/CMakeLists.txt [hunk 21]
```diff
@@ -336,20 +336,21 @@ if (BUILD_TESTS)
...
-    target_link_libraries(tool.drcacheoff.raw2trace_io drdecode)
-    use_DynamoRIO_extension(tool.drcacheoff.raw2trace_io drcovlib_static)
-    # Because we're leveraging instru_online code we have to link with drutil:
-    use_DynamoRIO_extension(tool.drcacheoff.raw2trace_io drutil_static)
-
-    # FIXME i#2099: the weak symbol is not supported on Windows.
     if (UNIX)
+      # uses ptrace, which is unix-only
+      add_executable(tool.drcacheoff.raw2trace_io tests/raw2trace_io.cpp
+                  tracer/instru.cpp
```
Comment (derekbruening) [inline clients/drcachesim/CMakeLists.txt]: Indent should be just 2 here and line below

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+            if (WIFEXITED(status))
+                break;
+            user_regs_struct regs;
+            res = ptrace(PTRACE_GETREGS, pid, NULL, &regs);
+            if (res < 0) {
+                perror("ptrace failed");
+                return 1;
+            }
+            /* We don't distinguish syscall entry/exit because orig_rax is set
+             * on both. */
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: style: prefer */ on own line

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+            if (SYS_NUM(regs) == __NR_open || SYS_NUM(regs) == __NR_openat ||
+                SYS_NUM(regs) == __NR_creat) {
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: In cmake `UNIX` includes MacOS where there is no __NR_creat.  If it's simpler you could just exclude in cmake (and ifdef sanity check at top of this file) via `NOT APPLE`.

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+        std::cerr << "Usage error: " << parse_err << "\nUsage:\n"
+                  << droption_parser_t::usage_short(DROPTION_SCOPE_ALL);
+        return 1;
+    }
+
+    /* Open input/output files outside of traced region. And explicitly don't destroy dir,
+     * so they never get closed.
+     */
+    raw2trace_directory_t *dir =
+        new raw2trace_directory_t(op_indir.get_value(), op_out.get_value());
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: Hmm, ok, it's not quite what I was thinking of (which was more of a test that actually uses the new i/o interface), but it does test the i/o isolation.

File: clients/drcachesim/CMakeLists.txt (hunk 21 above)
Comment (s-kanev) [inline clients/drcachesim/CMakeLists.txt]: Done.

File: clients/drcachesim/tests/raw2trace_io.cpp (hunk 24 above)
Comment (s-kanev) [inline clients/drcachesim/tests/raw2trace_io.cpp]: Ack.

File: clients/drcachesim/tests/raw2trace_io.cpp (hunk 22 above)
Comment (s-kanev) [inline clients/drcachesim/tests/raw2trace_io.cpp]: Done.

File: clients/drcachesim/tests/raw2trace_io.cpp (hunk 23 above)
Comment (s-kanev) [inline clients/drcachesim/tests/raw2trace_io.cpp]: Done.

**Comment by s-kanev** [comment]
@derekbruening, can you merge this? The CI failure looks completely unrelated -- I'm assuming from a flaky test.

**Comment by derekbruening** [comment]
OK, the general policy is to cite the issue for any flaky test or file one if it doesn't exist.  That looks like #1308.  This needs updating to master anyway so it will re-run.

**Pull request merged**